### PR TITLE
tool to generate a helm chart from an ACM MCE release

### PR DESCRIPTION
**Pull request by geoberle**
### What this PR does

extracts and templateizes the manifests from an MCE OLM release bundle

https://issues.redhat.com/browse/ARO-10315

this is part 1 or 2 - this part introduces the base repackage tool 
part 2 will use it to replace the current ACM installation (with OLM) to one with helm (without OLM)

Jira: 
Link to demo recording:

### CODE REVIEW

File: tooling/mcerepkg/internal/customize/customize.go [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+			for k, v := range newParams {
+				parameters[k] = v
+			}
+		}
+		customizedManifests[i] = obj
+	}
+	return customizedManifests, makeNestedMap(parameters), nil
+}
+
+func parameterizeNamespace(obj unstructured.Unstructured) (unstructured.Unstructured, map[string]string, error) {
```
[the remark this answers is not in the document]
Comment (geoberle) [reply tooling/mcerepkg/internal/customize/customize.go:61]: it needs to adhere to the `Customizer` type so all of them can be collected in the `customizerFuncs` list

File: tooling/mcerepkg/internal/customize/customize.go [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+		for c, container := range deployment.Spec.Template.Spec.Containers {
+			deployment.Spec.Template.Spec.Containers[c].Image = parameterizeImageRegistry(container.Image, imageRegistryParamName)
+		}
+		modifiedObj, err := convertToUnstructured(deployment)
+		return modifiedObj, map[string]string{imageRegistryParamName: ""}, err
+	}
+	return obj, nil, nil
+}
+
+func annotationCleaner(obj unstructured.Unstructured) (unstructured.Unstructured, map[string]string, error) {
```
[the remark this answers is not in the document]
Comment (geoberle) [reply tooling/mcerepkg/internal/customize/customize.go:142]: see above

File: tooling/mcerepkg/internal/rukpak/util/util.go [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+// 	return infosToObjects(infos), nil
+// }
+
+// func infosToObjects(infos []*resource.Info) []client.Object {
+// 	objects := make([]client.Object, 0, len(infos))
+// 	for _, info := range infos {
+// 		clientObject := info.Object.(client.Object)
+// 		objects = append(objects, clientObject)
+// 	}
+// 	return objects
```
[the remark this answers is not in the document]
Comment (geoberle) [reply tooling/mcerepkg/internal/rukpak/util/util.go]: done

File: tooling/mcerepkg/main.go [hunk 4]
```diff
@@ -0,0 +1,132 @@
+package main
```
[the remark this answers is not in the document]
Comment (geoberle) [reply tooling/mcerepkg/main.go:1]: done

**Comment by geoberle** [comment]
> Please add tests, many of these files mutate or operator on internal data structures, which should be tested from the beginning on.

i've added some basic tests revolving around the structure of the expected YAMLs.
i've also added some sanity checks that are run before helm chart generation starts.

File: tooling/mcerepkg/main.go [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+	}
+}
+
+func buildChart(outputDir string, mceOlmBundle string, scaffoldDir string) error {
+	ctx := context.Background()
+
+	// load OLM bundle manifests
+	img, err := crane.Load(mceOlmBundle)
+	if err != nil {
+		return err
```
Comment (janboll) [inline tooling/mcerepkg/main.go:61]: Please wrap error to make debugging easier

File: tooling/mcerepkg/main.go [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+	}
+	olmManifests, reg, err := olm.ExtractOLMBundleImage(ctx, img)
+	if err != nil {
+		return err
```
Comment (janboll) [inline tooling/mcerepkg/main.go:65]: Please wrap error to make debugging easier

File: tooling/mcerepkg/main.go [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+	// sanity check manifests
+	err = customize.SanityCheck(olmManifests)
+	if err != nil {
+		return fmt.Errorf("failed sanity checks on manifests: %v", err)
+	}
+
+	// load scaffolding manifests
+	scaffoldManifests, err := customize.LoadScaffoldTemplates(scaffoldDir)
+	if err != nil {
+		return err
```
Comment (janboll) [inline tooling/mcerepkg/main.go:77]: Please wrap error to make debugging easier

File: tooling/mcerepkg/main.go [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+	}
+
+	// customize manifests
+	customizedManifests, values, err := customize.CustomizeManifests(append(olmManifests, scaffoldManifests...))
+	if err != nil {
+		return err
```
Comment (janboll) [inline tooling/mcerepkg/main.go:83]: Please wrap error to make debugging easier

File: tooling/mcerepkg/main.go [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+			Name: path,
+			Data: yamlData,
+		})
+	}
+	mceChart.Templates = chartFiles
+
+	// store chart
+	err = chartutil.SaveDir(mceChart, outputDir)
+	if err != nil {
+		return err
```
Comment (janboll) [inline tooling/mcerepkg/main.go:134]: Please wrap error to make debugging easier

File: tooling/mcerepkg/internal/customize/customize.go [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+var (
+	operandImageEnvVarPrefix = "OPERAND_IMAGE_"
+	imageRegistryParamName   = "imageRegistry"
+)
+
+type Customizer func(unstructured.Unstructured) (unstructured.Unstructured, map[string]string, error)
```
Comment (janboll) [inline tooling/mcerepkg/internal/customize/customize.go:17]: Is this type enforced from outside? Why is it not using pointers? Could save some memory using i.e.:
```
type Customizer func(*unstructured.Unstructured) (map[string]string, error)
```
and update the object in the method

File: tooling/mcerepkg/internal/customize/scaffold.go [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+func LoadScaffoldTemplates(scaffoldDir string) ([]unstructured.Unstructured, error) {
+	var manifests []unstructured.Unstructured
+	err := filepath.Walk(scaffoldDir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if !info.IsDir() && filepath.Ext(path) == ".yaml" {
```
Comment (janboll) [inline tooling/mcerepkg/internal/customize/scaffold.go:17]: whay about `.yml` ?

**Review by janboll — CHANGES_REQUESTED** [review]

File: tooling/mcerepkg/main.go (hunk 8 above)
Comment (geoberle) [reply tooling/mcerepkg/main.go]: done

File: tooling/mcerepkg/main.go (hunk 6 above)
Comment (geoberle) [reply tooling/mcerepkg/main.go]: done

File: tooling/mcerepkg/internal/customize/scaffold.go (hunk 11 above)
Comment (geoberle) [reply tooling/mcerepkg/internal/customize/scaffold.go]: done

**Review by janboll — APPROVED** [review]

**Pull request merged**